ToolBox/KrakenDataDownloader/KrakenDataDownloader.cs

```diff
@@ -0,0 +1,141 @@
+{
+    /// <summary>
+    /// Kraken Data Downloader class
+    /// </summary>
+    class KrakenDataDownloader : IDataDownloader
+    {
+        private const string UrlPrototype = @"https://api.kraken.com/0/public/Trades?pair={0}&since={1}";
+
+        /// <summary>
+        /// Get historical data enumerable for a trading pair, type and resolution given this start and end time (in UTC).
+        /// </summary>
+        /// <param name="symbol">Symbol for the data we're looking for.</param>
+        /// <param name="resolution">Resolution of the data request</param>
+        /// <param name="startUtc">Start time of the data in UTC</param>
+        /// <param name="endUtc">End time of the data in UTC</param>
+        /// <returns>Enumerable of base data for this symbol</returns>
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            if (endUtc < startUtc)
+            {
+                throw new ArgumentException("The end date must be greater or equal than the start date.");
+            }
+
+            if (resolution != Resolution.Tick)
+            {
+                throw new NotSupportedException("Only Tick Resolution is supported.");
+            }
+
+            var startUnixTime = ToUnixTime(startUtc) * 1000000000; // Multiply by 10^9 per Kraken API
+            var endUnixTime = ToUnixTime(endUtc) * 1000000000;
+            var url = string.Format(UrlPrototype, symbol.Value, startUnixTime);
+            List<List<string>> data;
+
+            using (var client = new WebClient())
+            {
+                var response = client.DownloadString(url);
+                dynamic result = JsonConvert.DeserializeObject<dynamic>(response);
+                if (result.error.Count != 0)
+                {
+                    throw new Exception("Error in Kraken API: " + result.error[0]);
+                }
+
+                data = result.result[symbol.Value].ToObject<List<List<string>>>();
+
+                var last = Convert.ToInt64(result.result.last);
+                while (last < endUnixTime)
+                {
+                    url = string.Format(UrlPrototype, symbol.Value, last);
+                    response = client.DownloadString(url);
+                    result = JsonConvert.DeserializeObject<dynamic>(response);
+
+                    var errorCount = 0;
+                    while (result.error.Count != 0 && errorCount < 10)
+                    {
+                        errorCount++;
+                        System.Threading.Thread.Sleep(6000);
+                        response = client.DownloadString(url);
+                        result = JsonConvert.DeserializeObject<dynamic>(response);
+                    }
+
+                    if (result.error.Count != 0 && errorCount >= 10)
+                    {
+                        throw new Exception("Error in Kraken API: " + result.error[0]);
```

Comment:
Should probably check if the error array; chance of out of bounds exception here

Reply:
What do you mean? result.error is an empty JSON array if the API call returned correctly. If we enter the if-statement, we have guaranteed that the size of the array >=1 (an error happened and Kraken is telling us why), so result.error[0] should never have an out-of-bounds exception. Am I reading this correctly?